### PR TITLE
Libraries: Language modeling with PyTorch (1)

### PR DESCRIPTION
**Pull request by edoardob90**
Part 1 notebook introduces the fundamentals of neural network-based language modeling, from traditional bi-gram approaches to the simplest neural networks.

- Implementation of a single-layer neural network for character-level language modeling
- Comparison with the bigram model approach, highlighting similarities in performance but differences in flexibility
- Step-by-step explanation of the neural network pipeline:
  - One-hot encoding of character inputs
  - Forward pass through a weight matrix
  - Softmax transformation to obtain probability distributions
  - Loss calculation using negative log-likelihood
  - Backward pass for gradient computation
  - Weight updates using gradient descent
- Introduction to regularization
- Demonstration of sampling from the trained model

It's the first step of a step-by-step introduction/overview of language modeling using PyTorch library.

### CODE REVIEW

**Comment by edoardob90** [comment]
**Left to do:**

- [ ] Add missing references
- [ ] Update Table of Contents
- [ ] Finalize the exercises

**Comment by edoardob90** [comment]
- [ ] Add solutions notebook

File: 32_language_modeling_1.ipynb [hunk 1]
```diff
@@ -0,0 +1,1958 @@
...
+    "from collections import defaultdict\n",
+    "\n",
+    "bigrams = defaultdict(int)\n",
+    "\n",
+    "for w in words:\n",
+    "    chars = [\"<S>\"] + list(w) + [\"<E>\"]\n",
+    "    for ch1, ch2 in zip(chars, chars[1:]):\n",
+    "        bigrams[(ch1, ch2)] += 1\n",
+    "\n",
+    "print(dict(bigrams))"
```
Comment (despadam) [inline 32_language_modeling_1.ipynb:205]: Why is this marked with `Student code`? What is the purpose of this?

Comment (edoardob90) [reply 32_language_modeling_1.ipynb:205]: Was a placeholder for myself: I'll remove them.

I thought I would delete some code to let them write it on their own, but I think I will suggest them to _copy_ the notebook and not simply scroll/read it through.

Comment (despadam) [reply 32_language_modeling_1.ipynb]: Ok, in general I think this is great work! And I would also suggest to guide them through it instead of just letting them read

**Comment by despadam** [comment]
Also, this should be included in `00_index.ipynb`

**Review by despadam — APPROVED** [review]
LGTM 👏

**Review by Snowwpanda — COMMENTED** [review]
Looks nice, good work.
